### PR TITLE
feat: handle access/delegate invocations without error

### PR DESCRIPTION
**Pull request by gobengo**
Note:
* this is a superset of https://github.com/web3-storage/w3protocol/pull/420
  * that defines the `access/delegate` capability parsers that this PR adds handlers for in access-api 
  * my intention is to land that #420 before this

Motivation:
* part of https://github.com/web3-storage/w3protocol/issues/425

### CODE REVIEW

File: packages/access-client/src/types.ts
Comment (gobengo) [inline packages/access-client/src/types.ts:94]: afaict there is no response type [in the spec](https://github.com/web3-storage/specs/blob/main/w3-access.md#accessdelegate).

It also looks like `void` is a fine `Success` placeholder for now, because I don't expect the invoker to need anything from  the `Success` other than knowing it wasn't a Failure. i.e. if `access/delegate` invocation was handled without error, then the delegations were persisted.

**Comment by gobengo** [comment]
@Gozala I added [this](https://github.com/web3-storage/w3protocol/pull/427/commits/e0834964e6a2d0ab2f5bcdd012c45c9c02dda80a) to make it so this patch is safe to merge/release even while I still intend to add better validation/hardening before invoking `access/delegate` should actually store delegations in the db in prod, which I will probably need to iterate on a bit after I'm out next week.

I'm hoping with this change, I can get some early review feedback now, and if there's nothing too scary, we could even consider merging this partial progress into main, so others can build on it while I'm out, but there will be no risk of half-baked functionality running in production if main branch is released to production.

File: packages/access-api/src/service/access-claim.js
Comment (gobengo) [inline packages/access-api/src/service/access-claim.js:47]: I will resolve this todo as part of https://github.com/web3-storage/w3protocol/issues/394
For now this implementation is sufficient for the most basic of 'test delegate, then claim' that I started.
But as that got larger, I realized I should just focus on `access/delegate` in this review round.

**Comment by gobengo** [comment]
@alanshaw wdyt? It's a lot of new code and a MVP `access/delegate`, but has feature switches so the new functionality won't be enabled in production. I'd love some feedback. If it looks safe, perhaps we could merge it into main to unblock further test/dev against staging.

File: packages/access-api/src/models/spaces.js
Comment (alanshaw) [inline packages/access-api/src/models/spaces.js:25]: Why "EMPTY" and not just ""?

File: packages/access-api/src/bindings.d.ts
Comment (alanshaw) [inline packages/access-api/src/bindings.d.ts:63]: Please change all or be consistent...🙏

```suggestion
    delegations: Delegations
```

File: packages/access-api/src/models/delegations.js
Comment (alanshaw) [inline packages/access-api/src/models/delegations.js:40]: It's unusual (IMHO) for a property accessor to return a promise. Why not a function?

```suggestion
  async length() {
```

File: packages/access-api/src/models/delegations.js
Comment (alanshaw) [inline packages/access-api/src/models/delegations.js:66]: Many questions:

* Why avoiding `await`?
* Why not return a `BigInt`?

```suggestion
    const data = await this.#db
        .selectFrom('delegations')
        .select((e) => e.fn.count('cid').as('size'))
        .executeTakeFirstOrThrow()
    return BigInt(data.size)
```

File: packages/access-api/src/models/delegations.js
Comment (alanshaw) [inline packages/access-api/src/models/delegations.js:75]:
```suggestion
  async push (...delegations) {
```

File: packages/access-api/src/models/delegations.js
Comment (alanshaw) [inline packages/access-api/src/models/delegations.js:93]: We should not allow this if the database cannot stream.

File: packages/access-api/src/service/access-delegate.js
Comment (alanshaw) [inline packages/access-api/src/service/access-delegate.js:13]: Can we subclass Ucanto's [`Failure`](https://github.com/web3-storage/ucanto/blob/main/packages/validator/src/error.js#L8) instead?

File: packages/access-api/src/service/access-delegate.js
Comment (alanshaw) [inline packages/access-api/src/service/access-delegate.js:84]: Maybe personal preference but I'd prefer collecting in an array to using a generator unless there's some kind of streaming benefit to be had.

File: packages/access-api/src/service/access-delegate.js
Comment (alanshaw) [inline packages/access-api/src/service/access-delegate.js:69]: Is more `MissingProvider`...no?

File: packages/access-api/src/types/delegations.ts
Comment (alanshaw) [inline packages/access-api/src/types/delegations.ts:13]: I'm not convinced this is a `push` - it's not ordered in any way - `put` may be more appropriate or `putMany`

File: packages/access-api/src/types/delegations.ts
Comment (alanshaw) [inline packages/access-api/src/types/delegations.ts:3]: Do we need this right now? Is `Promise<T>` ok? Commonly in nodejs you'd find `*Sync` suffixed versions.

**Review by alanshaw — CHANGES_REQUESTED** [review]

File: packages/access-api/src/types/delegations.ts
Comment (Gozala) [reply packages/access-api/src/types/delegations.ts:3]: We have `Await` type in multiformats for this and I think it’s also used and reexported by ucanto

**Review by Gozala — COMMENTED** [review]
I really quickly glanced at the pr and noticed that we’m be pulling delegation as bytes from db, decode them into delegations which then ucanto will take and encode back into bytes. Decode/encode there seems pretty wasteful & perhaps can be avoided by sending bytes and let client decode

File: packages/access-api/src/models/spaces.js
Comment (gobengo) [reply packages/access-api/src/models/spaces.js:27]: https://github.com/web3-storage/w3protocol/issues/447

Comment (gobengo) [reply packages/access-api/src/models/spaces.js:27]: I pushed a comment explaining the magic `EMPTY`

File: packages/access-api/src/bindings.d.ts
Comment (gobengo) [reply packages/access-api/src/bindings.d.ts]: I changed the import name so it looks like the others.

The thing is, this isn't the same as the others. This is an interface, whereas the other 'models' here point to classes.
@Gozala and I discussed in this work trying eager to apply the [dependency inversion principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle) and depend on an interface here instead of an implementation/class.

So that's why the name is inconsistent, it's not doing the same thing as others.

File: packages/access-api/src/models/delegations.js
Comment (gobengo) [reply packages/access-api/src/models/delegations.js:40]: originally this type started as a subset of `Array` and I did some TDD from there to implement the interface backed by sql, and that required affording for async. I figured it would make sense to adjust the names/types later in the context of a specific reviewer. Happy to change this to a function now. will do

Comment (gobengo) [reply packages/access-api/src/models/delegations.js:40]: IMHO properties make sense for nouns like 'length' and functions make sense for methods. I'm gonna call the function count() (inspired by sql count(), mongo count(), etc). can change it again if you def want 'length()'

File: packages/access-api/src/types/delegations.ts
Comment (gobengo) [reply packages/access-api/src/types/delegations.ts:3]: We don't *need* it. But right now an `Array` satisfies the `DelegationsStorage` type and getting rid of it would remove that property which, for sure, we don't need. (I started with an `Array` and expanded the type from there to afford for async and adding `Resolvable` was an easy way of doing that)

File: packages/access-api/src/models/delegations.js
Comment (gobengo) [reply packages/access-api/src/models/delegations.js:66]: * couldn't do `await` when `length` was a property. but now its not so i'll adjust.
* will also make it a bigint native now that I'm making DelegationsStorage not be met by an `Array`

File: packages/access-api/src/models/delegations.js
Comment (gobengo) [reply packages/access-api/src/models/delegations.js]: renamed to `count(): Promise<bigint>` in https://github.com/web3-storage/w3protocol/pull/427/commits/022dc2167647dcbabb35b6ab0450dafd9c708a84

File: packages/access-api/src/models/delegations.js
Comment (gobengo) [reply packages/access-api/src/models/delegations.js]: made those changes in https://github.com/web3-storage/w3protocol/pull/427/commits/022dc2167647dcbabb35b6ab0450dafd9c708a84

File: packages/access-api/src/models/delegations.js
Comment (gobengo) [reply packages/access-api/src/models/delegations.js]: made this change in . https://github.com/web3-storage/w3protocol/pull/427/commits/022dc2167647dcbabb35b6ab0450dafd9c708a84

File: packages/access-api/src/models/delegations.js
Comment (gobengo) [reply packages/access-api/src/models/delegations.js]: Why not? Would it be sufficient to e.g. only refuse to do that in production or if the overall db size is very large.

I removed the non-streaming case for now, but it was needed for a WIP version of the `testDelegateThenClaim` test. https://github.com/web3-storage/w3protocol/actions/runs/4256209423/jobs/7404788157#step:7:48
But for now I've removed it and will just disable tests that relied on it (since further work #394 will lead to a better impl that doesn't need asyncIteration).

File: packages/access-api/src/service/access-delegate.js
Comment (gobengo) [reply packages/access-api/src/service/access-delegate.js:13]: Why is that preferable to defining a type like this? This type structurally matches to `Ucanto.Failure`, and if it deviates, the type-checker will catch that.

A reason I would avoid subclassing here is that subclassing would be a concrete implementation that depended via prototype chain on a specific instance/version of `Ucanto.Failure`. I intentionally want to avoid that dependency on the entirety of `Ucanto.Failure`. A lot of generic methods in ucanto ecosystem operate on type `{ error: true }` or `API.Failure` from `@ucanto/interface` but that specific `Failure` class you implement also has e.g. a `describe` method that I don't want/need here. I avoided subclassing in order to apply https://en.wikipedia.org/wiki/Interface_segregation_principle

File: packages/access-api/src/service/access-delegate.js
Comment (gobengo) [reply packages/access-api/src/service/access-delegate.js:84]: defining this as a generator function yields delegations as we enounter them, which affords for streaming them out.

You can trivially collect them into an array like:
```
const collected = Array.from(extractProvenDelegations(invocation))
```

but if I authored this logic as a non-generator, it wouldn't be as useful to anyone that wanted to stream or e.g. only extract the first N proven delegation.

File: packages/access-api/src/service/access-delegate.js
Comment (gobengo) [reply packages/access-api/src/service/access-delegate.js:69]: There is no notion in the protocols implemented today about a storage provider, so I want to avoid mentioning that concept and committing to it in the API indefinitely https://github.com/web3-storage/w3protocol/search?q=provider

But part of the issue here is that the [protocol spec](https://github.com/web3-storage/specs/blob/main/w3-access.md#accessdelegate) doesn't define the error types. I don't think we necessarily should add error names for all of the distinct things that could happen in our implementation here, so much as add just enough error types as are needed to make the protocol useful to a client regardless of implementation. I like `InsufficientStorage` because it's more than enough here to let the user know that retrying the request wont work, and it's also vague enough that we can use it for other situations. That was why I did it the way I did, but I could have painted the bikeshed another color. 

@alanshaw Can you document a use case for needing to distinguish `MissingProvider` from `InsufficientStorage`?

@Gozala I know sometimes you have success/failure types in mind in your specs, and sometimes even can point to earlier implementations where you specified some. Do you have any thoughts on what the `Failure` type for `access/delegate` should be?

We can also iterate on the details of this without blocking this PR, since this is all new functionality and the PR as written now disables the functionality in prod, so we can make backward-incompatible changes

File: packages/access-api/src/types/delegations.ts
Comment (gobengo) [reply packages/access-api/src/types/delegations.ts]: Removed `Resolvable` in https://github.com/web3-storage/w3protocol/pull/427/commits/022dc2167647dcbabb35b6ab0450dafd9c708a84#diff-0c692d0f8036f645c9fe7c647d216265dedfb7a84decb52490918e43750a9e9eL3

File: packages/access-api/src/types/delegations.ts
Comment (gobengo) [reply packages/access-api/src/types/delegations.ts]: renamed to `putMany` https://github.com/web3-storage/w3protocol/pull/427/commits/022dc2167647dcbabb35b6ab0450dafd9c708a84#diff-0c692d0f8036f645c9fe7c647d216265dedfb7a84decb52490918e43750a9e9eR11

**Comment by gobengo** [comment]
> I really quickly glanced at the pr and noticed that we’ll be pulling delegation as bytes from db, decode them into delegations which then ucanto will take and encode back into bytes. Decode/encode there seems pretty wasteful & perhaps can be avoided by sending bytes and let client decode

This is a good point and something we can optimize later. I think this only matters in the context of `access/claim` which I stubbed out here but plan to flesh out after this in #394 . I added some WIP `access/claim`-related code in here, but none of it is reachable from the `access-api` worker because the underlying service only adds `access.delegate` and not `access.claim`, so I don't think merging this will lead to the performance concern *in the wild*, so I suggest that it is an acceptable risk to handle later as part of #394

File: packages/access-api/src/service/access-delegate.js
Comment (alanshaw) [reply packages/access-api/src/service/access-delegate.js:13]: > Why is that preferable to defining a type like this?

There is more than one place where we'll want this error - I think it will be less code overall to just import and use.

> I intentionally want to avoid that dependency on the entirety of that class.

It is tiny, and re-exported from `@ucanto/server` anyway which you're already including here.

File: packages/access-api/src/service/access-delegate.js
Comment (alanshaw) [reply packages/access-api/src/service/access-delegate.js:84]: I don't think I saw a significant amount of work that the function was doing that would benefit from it being done incrementally.

When this function is being used, you immediately spread the results into a function call, which is not using the generator in a streaming fashion anyway.

If in the future you need to consume part of the stream, or it is blocking the event loop or you want to do some work between recieving each proof then sure, make it a generator. Right now though, this function reads to _me_ as though it does some heavy lifting or as a consumer I'm going to want/need to process the data incrementally, which doesn't appear to be the case.

File: packages/access-api/src/service/access-delegate.js
Comment (alanshaw) [reply packages/access-api/src/service/access-delegate.js:69]: No feels the same, happy to go with `InsufficientStorage` for now.

File: packages/access-api/src/service/access-delegate.js
Comment (alanshaw) [reply packages/access-api/src/service/access-delegate.js:84]: This is not blocking!

File: packages/access-api/src/service/access-delegate.js
Comment (alanshaw) [reply packages/access-api/src/service/access-delegate.js:13]: This is not blocking!

File: packages/access-api/src/models/delegations.js
Comment (alanshaw) [reply packages/access-api/src/models/delegations.js]: Hmm so at the time I didn't know that this was going to become a filtered list - in the case of all data then there's no point in allowing buffered response as it'll immediately become unusable.

For a specific DID's delegations then sure buffered response is fine.

File: packages/access-api/src/bindings.d.ts
Comment (alanshaw) [reply packages/access-api/src/bindings.d.ts]: OK gotcha - next time just add that info to the PR description so I know it's intentional!

**Review by alanshaw — APPROVED** [review]

File: packages/access-api/src/service/access-delegate.js
Comment (gobengo) [reply packages/access-api/src/service/access-delegate.js:13]: > There is more than one place where we'll want this error - I think it will be less code overall to just import and use.

to reuse, import the type from `import('@web3-storage/capabilities/types').InsufficientStorage}` as is done here

I don't need a subclass or really any class here, so I'm not going to add one, but if anyone else wants to, they can use that import with `@implements` to ensure it will jive with everything here.

**Pull request merged**
